src/main/scala/beam/agentsim/infrastructure/parking/ParkingStallSampling.scala

```diff
@@ -3,15 +3,100 @@ package beam.agentsim.infrastructure.parking
 import scala.util.Random
 import beam.agentsim.infrastructure.taz.TAZ
 import beam.router.BeamRouter.Location
-import org.matsim.api.core.v01.Coord
+import beam.utils.logging.ExponentialLazyLogging
+import org.matsim.api.core.v01.network.Link
+import org.matsim.api.core.v01.{Coord, Id}
+import org.matsim.core.utils.collections.QuadTree
+
+import scala.jdk.CollectionConverters.collectionAsScalaIterableConverter
+import scala.math.pow
 
 /**
   * sampling methods for randomly generating stall locations from aggregate information
   */
-object ParkingStallSampling {
+object ParkingStallSampling extends ExponentialLazyLogging {
 
   val maxOffsetDistance = 600.0 // TODO: Make this a config parameter
 
+  def linkBasedSampling(
+    rand: Random,
+    requestLocation: Location,
+    maybeLinkQuadTree: Option[QuadTree[Link]],
+    distanceFunction: (Coord, Coord) => Double,
+    availabilityRatio: Double,
+    taz: TAZ,
+    inClosestZone: Boolean,
+    maxDist: Double = maxOffsetDistance
+  ): Location = {
+    maybeLinkQuadTree match {
+      case Some(linkQuadTree) =>
+        val allLinks = linkQuadTree.getDisk(requestLocation.getX, requestLocation.getY, maxDist).asScala
+        val totalLength = allLinks.foldRight(0.0)(_.getLength + _)
+        var currentLength = 0.0
+        val filteredLinks = rand.shuffle(allLinks).takeWhile { lnk =>
+          currentLength += lnk.getLength
+          currentLength <= totalLength * availabilityRatio
+        }
+        Some(filteredLinks)
+          .filter(_.nonEmpty)
```

Comment:
This can be pushed out and filtered earlier.

Reply:
In fact, this is the only reason you would do a `getOrElse` as near as I can tell. I think an if/else would be fine here:

```
if(filteredLinks.isEmpty) //logger.warn....etc
else //filteredLinks.map....etc
```